samples/StaticFilesAuth/Startup.cs

```diff
@@ -114,81 +138,61 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env, IAuthori
+            var fileSystemInfo = GetFileSystemInfo(files, filePath);
+            if (fileSystemInfo != null)
             {
-                var fileInfo = files.GetFileInfo(context.Request.Path);
-                AuthorizationResult result = null;
-                if (fileInfo.Exists)
-                {
-                    result = await authorizationService.AuthorizeAsync(context.User, fileInfo, "files");
-                }
-                else
-                {
-                    // https://github.com/aspnet/Home/issues/2537
-                    var dir = files.GetDirectoryContents(context.Request.Path);
-                    if (dir.Exists)
-                    {
-                        result = await authorizationService.AuthorizeAsync(context.User, dir, "files");
-                    }
-                    else
-                    {
-                        context.Response.StatusCode = StatusCodes.Status404NotFound;
-                        return;
-                    }
-                }
+                var metadata = new List<object>();
+                metadata.Add(fileSystemInfo);
+                metadata.Add(new AuthorizeAttribute(policy));
+
+                var endpoint = new Endpoint(
+                    c => Task.CompletedTask,
+                    new EndpointMetadataCollection(metadata),
+                    context.Request.Path);
+
+                context.Features.Set<IEndpointFeature>(new EndpointFeature(endpoint));
+            }
+        }
 
-                if (!result.Succeeded)
+        private static object GetFileSystemInfo(PhysicalFileProvider files, string path)
+        {
+            var fileInfo = files.GetFileInfo(path);
+            if (fileInfo.Exists)
+            {
+                return fileInfo;
+            }
+            else
+            {
+                // https://github.com/aspnet/Home/issues/2537
+                var dir = files.GetDirectoryContents(path);
+                if (dir.Exists)
                 {
-                    if (!context.User.Identity.IsAuthenticated)
-                    {
-                        await context.ChallengeAsync(new AuthenticationProperties()
-                        {
-                            // https://github.com/aspnet/Security/issues/1730
-                            // Return here after authenticating
-                            RedirectUri = context.Request.PathBase + context.Request.Path + context.Request.QueryString
-                        });
-                        return;
-                    }
-                    // Authenticated but not authorized
-                    await context.ForbidAsync();
-                    return;
+                    return dir;
                 }
+            }
 
-                await next();
-            });
-            branch.UseFileServer(new FileServerOptions()
+            return null;
+        }
+
+        private class EndpointFeature : IEndpointFeature
```

Comment:
Callers shouldn't have to create this dummy type. See the comments about Get/SetEndpoint() extensions.